File: clsMercuryIsotopeDistribution.cpp
// Written by Navdeep Jaitly for the Department of Energy (PNNL, Richland, WA)
// Copyright 2006, Battelle Memorial Institute
// E-mail: navdeep.jaitly@pnl.gov
// Website: http://ncrr.pnl.gov/software
// -------------------------------------------------------------------------------
// 
// Licensed under the Apache License, Version 2.0; you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at 
// http://www.apache.org/licenses/LICENSE-2.0

#pragma once
#include "clsMercuryIsotopeDistribution.h"
#include "DeconEngineUtils.h"

namespace DeconToolsV2
{
	clsMercuryIsotopeDistribution::clsMercuryIsotopeDistribution()
	{
		mMercuryIsotopeDistribution = new Engine::TheoreticalProfile::MercuryIsotopeDistribution();
		MercurySize(8192);
		ChargeCarrierMass(1.007824659); 
		ApodizationType(ApodizationType::Gaussian);
		Resolution(100000);
		ChargeState(1);
	}

	clsMercuryIsotopeDistribution::~clsMercuryIsotopeDistribution()
	{
		if (mMercuryIsotopeDistribution != NULL)
		{
			delete mMercuryIsotopeDistribution;
			mMercuryIsotopeDistribution = NULL ; 
		}
	}


	array<System::Drawing::PointF^>^ clsMercuryIsotopeDistribution::CalculateDistribution(System::Collections::Hashtable^ elementCounts)
	{
		std::vector<double> vect_x;
		std::vector<double> vect_y;

		std::vector<double> vect_isotope_x;
		std::vector<double> vect_isotope_y;

		array<System::Drawing::PointF^>^ points;

		Engine::TheoreticalProfile::MolecularFormula formula ; 
		DeconEngine::Utils::ConvertElementTableToFormula(mMercuryIsotopeDistribution->mobj_elemental_isotope_composition, elementCounts, formula);

		mMercuryIsotopeDistribution->CalculateDistribution(mChargeState, mResolution, formula, vect_x, vect_y, 0, vect_isotope_x, vect_isotope_y, false);

		int num_pts = (int) vect_x.size();
		points = gcnew array<System::Drawing::PointF^>(num_pts);

		for (int i = 0 ; i < num_pts ; i++)
		{
			points[i] = System::Drawing::PointF((float) vect_x[i], (float) vect_y[i]);
		}

		return points;
	};

	void clsMercuryIsotopeDistribution::CalculateMasses(System::Collections::Hashtable^ elementCounts) 
	{		
		Engine::TheoreticalProfile::MolecularFormula formula ; 
		DeconEngine::Utils::ConvertElementTableToFormula(mMercuryIsotopeDistribution->mobj_elemental_isotope_composition, elementCounts, formula);

		mMercuryIsotopeDistribution->CalculateMasses(formula);
	}
}

File: C#_Version/DeconEngine/clsMercuryIsotopeDistribution.cs
using System;
using System.Collections;
using System.Drawing;
using DeconEngine;
using Engine.TheoreticalProfile;

namespace DeconToolsV2
{
    [Obsolete("Only used by Decon2LS.UI", false)]
    public class clsMercuryIsotopeDistribution
    {
        private int mChargeState;
        private readonly MercuryIsotopeDistribution mMercuryIsotopeDistribution;

        public clsMercuryIsotopeDistribution()
        {
            mMercuryIsotopeDistribution = new MercuryIsotopeDistribution();
            MercurySize = 8192;
            ChargeCarrierMass = 1.007824659;
            ApodizationType = ApodizationType.Gaussian;
            Resolution = 100000;
            ChargeState = 1;
        }

        /// <summary>
        /// Gets or Sets the mercury size. Only accepts multiples of 2.
        /// </summary>
        public int MercurySize
        {
            get => mMercuryIsotopeDistribution.MercurySize;
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("value", value,
                        "Mercury size must be greater than 0");
                }

                var goodSize = false;
                for (var i = 1; i <= 536870912; i *= 2)
                {
                    if (value == i)
                    {
                        goodSize = true;
                        break;
                    }
                }
                if (!goodSize)
                {
                    throw new ArgumentException("MercurySize must be a power of two.", "value");
                }

                mMercuryIsotopeDistribution.MercurySize = value;
            }
        }

        public double ChargeCarrierMass
        {
            get => mMercuryIsotopeDistribution.ChargeCarrierMass;
            set => mMercuryIsotopeDistribution.ChargeCarrierMass = value;
        }

        public ApodizationType ApodizationType
        {
            get => mMercuryIsotopeDistribution.ApType;
            set => mMercuryIsotopeDistribution.ApType = value;
        }

        public double Resolution { get; set; }

        /// <summary>
        /// Gets and sets charge state property
        /// </summary>
        public short ChargeState
        {
            get => (short) mChargeState;

            set => mChargeState = value;
        }

        /// <summary>
        /// Gets average molecular weight of the last call to CalculateDistribution
        /// </summary>
        public double AverageMolecularMass => mMercuryIsotopeDistribution.AverageMw;

        /// <summary>
        /// Gets the monoisotopic molecular weight of the last call to CalculateDistribution
        /// </summary>
        public double MonoMolecularMass => mMercuryIsotopeDistribution.MonoMw;

        /// <summary>
        /// Gets the most abundant MZ of the last call to CalculateDistribution
        /// </summary>
        public double MostAbundantMZ => mMercuryIsotopeDistribution.MaxPeakMz;

        /// <summary>
        /// Gets the mass variance of the last call to CalculateDistribution
        /// </summary>
        public double MassVariance => mMercuryIsotopeDistribution.MassVariance;

        public clsElementIsotopes ElementIsotopes
        {
            set => mMercuryIsotopeDistribution.SetElementalIsotopeComposition(
                value);
        }

        /// <summary>
        /// Call the mMercuryIsotopeDistribution method, passing in the currently set resolution and other settings.  Copy the
        /// result into an array of PointF.
        /// </summary>
        /// <param name="elementCounts"></param>
        /// <returns></returns>
        public PointF[] CalculateDistribution(Hashtable elementCounts)
        {
            Utils.ConvertElementTableToFormula(
                ref mMercuryIsotopeDistribution.ElementalIsotopeComposition, elementCounts, out var formula);

            mMercuryIsotopeDistribution.CalculateDistribution((short) mChargeState, Resolution, formula, out var x, out var y, 0,
                out var xIsotope, out var yIsotope, false);

            var numPts = x.Count;
            var points = new PointF[numPts];

            for (var i = 0; i < numPts; i++)
            {
                points[i] = new PointF((float) x[i], (float) y[i]);
            }

            return points;
        }

        public void CalculateMasses(Hashtable elementCounts)
        {
            Utils.ConvertElementTableToFormula(
                ref mMercuryIsotopeDistribution.ElementalIsotopeComposition, elementCounts, out var formula);

            mMercuryIsotopeDistribution.CalculateMasses(formula);
        }
    }
}
